library/std/src/net/tcp/tests.rs

```diff
@@ -46,6 +46,22 @@ fn connect_error() {
     }
 }
 
+#[test]
+fn connect_timeout_error() {
+    match TcpStream::connect_timeout(&"0.0.0.0:1".parse().unwrap(), Duration::MAX) {
```

Comment:
Have you confirmed this test fails without the changes in `net.rs`?

Reply:
@eval-exec for an alternative idea of what a test could do:

1. bind some free port on localhost using `TcpListener::bind`, then
2. connect to that new port with `Duration::MAX` timeout.

Here's a draft in Rust:

```rust
use std::net::TcpListener;

let listener = TcpListener::bind("127.0.0.1:0").unwrap();  // :0 picks some free port
let port = listener.local_addr().unwrap().port();  // obtain the port it picked

[..]

[..] format!("127.0.0.1:{port}").as_str() [..]  // connect to that port
```

I have a real example of something very close at https://github.com/hartwork/rust-for-it/blob/4e51982c4465c3190d92a59747c5d2738baf9e07/src/network.rs#L160-L177 .

What do you think?

Reply:
Thank you, I'm testing it on a win10 virtual machine.

Reply:
@eval-exec I checked the latest version: `0.0.0.0:1` will only work if you start a TCP listener on port 1. And port 1 only works if its unoccupied and you have privileges because ports up to 1024 are reserved. In a test context you'll probably want a dynamic free port as in my example above.

Reply:
I'm confused, I try to connect to an unreachable address `8.8.8.8:9999`, but it didn't get timeout.

Reply:
I want to follow the test here: https://github.com/bnoordhuis/libuv-chat/blob/master/deps/uv/test/test-tcp-connect-timeout.c

Reply:
@eval-exec given that the port is connectable now, the expected result is `Ok(..)` and so I would just assert for `.is_ok()`and all the matching can be dropped and you'll have only three lines for a test body. What do you think?

Reply:
I execute `telnet 8.8.8.8 9999` but did get timeout, because I'm running clash with TUN mode.

after I stop clash TUN mode, `telnet 8.8.8.8 9999` will get timeout.

Reply:
@hartwork Thank you, I think I didn't fully understand you, could you please provide the code?

Reply:
> @hartwork Thank you, I think I didn't fully understand you, could you please provide the code?

@eval-exec this is the adjustment that I had in mind:

```rust
use std::net::{TcpListener, TcpStream};
use std::time::Duration;

fn main() {
    let listener = TcpListener::bind("127.0.0.1:0").unwrap(); // :0 picks some free port
    let port = listener.local_addr().unwrap().port(); // obtain the port it picked
    assert!(
        TcpStream::connect_timeout(&format!("0.0.0.0:{port}").parse().unwrap(), Duration::MAX)
            .is_ok()
    );
}
```

Reply:
@eval-exec updated twice more (in case you only saw the version sent via notification e-mail)

Reply:
It should be `format!("0.0.0.0:{port}"` or `format!("127.0.0.1:{port}"` for `connect_timeout`?

Reply:
I think I should `bind` and `accept` in another thread.

Reply:
> I think I should `bind` and `accept` in another thread:

@eval-exec that's not necessary, see https://github.com/hartwork/rust-for-it/blob/4e51982c4465c3190d92a59747c5d2738baf9e07/src/network.rs#L160-L177 .